server/middleware.ts

```diff
@@ -0,0 +1,201 @@
+		});
+	}
+	else {
+		next();
+	}
+}
+
+// For API endpoints
+export function authenticateWithReject(request: express.Request, response: express.Response, next: express.NextFunction) {
+	if (!request.isAuthenticated()) {
+		response.status(401).json({
+			"error": "You must log in to access this endpoint"
+		});
+	}
+	else {
+		next();
+	}
+}
+// For directly user facing endpoints
+export function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
+	if (!request.isAuthenticated()) {
+		response.redirect("/login");
+	}
+	else {
+		next();
+	}
+}
+import * as Handlebars from "handlebars";
+import { ICommonTemplate } from "./schema";
+export enum ApplicationType {
+	Application, Confirmation
+}
+export async function timeLimited(request: express.Request, response: express.Response, next: express.NextFunction) {
+	let requestType: ApplicationType = request.url.match(/^\/apply/) ? ApplicationType.Application : ApplicationType.Confirmation;
+
+	let openBranches: (ApplicationBranch | ConfirmationBranch)[];
+	if (requestType === ApplicationType.Application) {
+		openBranches = await BranchConfig.getOpenBranches<ApplicationBranch>("Application");
+	}
+	else {
+		openBranches = await getOpenConfirmationBranches(request.user as IUser);
+	}
+
+	if (openBranches.length > 0) {
+		next();
+		return;
+	}
+
+	// TODO reimplement open and close times
+	/*
+	const TIME_FORMAT = "dddd, MMMM Do YYYY [at] h:mm a z";
+	*/
+	interface IClosedTemplate extends ICommonTemplate {
+		type: string;
+		/*
+		open: {
+			time: string;
+			verb: string;
+		};
+		close: {
+			time: string;
+			verb: string;
+		};
+		*/
+		contactEmail: string;
+	}
+	let template = Handlebars.compile(await readFileAsync(path.resolve(STATIC_ROOT, "closed.html")));
+	let emailParsed = config.email.from.match(/<(.*?)>/);
+	let templateData: IClosedTemplate = {
+		siteTitle: config.eventName,
+		user: request.user,
+		settings: {
+			teamsEnabled: await getSetting<boolean>("teamsEnabled"),
+			qrEnabled: await getSetting<boolean>("qrEnabled")
+		},
+
+		type: requestType === ApplicationType.Application ? "Application" : "Confirmation",
+		/*
+		open: {
+			time: openDate.tz(moment.tz.guess()).format(TIME_FORMAT),
+			verb: moment().isBefore(openDate) ? "will open" : "opened"
+		},
+		close: {
+			time: closeDate.tz(moment.tz.guess()).format(TIME_FORMAT),
+			verb: moment().isBefore(closeDate) ? "will close" : "closed"
+		},
+		*/
+		contactEmail: emailParsed ? emailParsed[1] : config.email.from
+	};
+	response.send(template(templateData));
+}
+
+import { DataLog, HackGTMetrics } from "./schema";
+export function trackEvent(action: string, request: express.Request, user?: string, data?: object) {
+	let thisEvent: DataLog = {
+		action,
+		url: request.path,
+		time: moment.utc().format(),
+		ip: request.ip,
+		user,
+		userAgent: request.headers["user-agent"] as string
+	};
+	console.log(thisEvent);
```

Comment:
log `JSON.stringify(thisEvent)` to escape possible newlines in the string. @ehsanmasdar is this correct?

Reply:
@illegalprime Not that necessary since the event is logged for HackGMetrics purposes below.